### ISSUE TITLE
Depending on how you define the function, typecheck is behaving differently even if is the same.

### ISSUE THREAD

**Issue by stevemarksd** [microsoft/TypeScript#35774]
**TypeScript Version:**  3.7.2

**Code**

```ts
const a1: () => { a: string } = () => {
  return {a: '', b: ''} // not erroring!?
}

const a2 = ():{ a: string } => {
  return {a: '', b: ''} // is erroring as expected
}

const a3: () => { a: string } = () => {
  return {b: ''} // is erroring as expected
}
```

**Expected behavior:**

a1 should error

**Actual behavior:**

a2 doesnt error

**Playground Link:** https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=11&pc=2#code/MYewdgzgLgBAhgRgFwwBQEoYF4B8MDe8K0ATgJZgDmMAvtmprgQFAwwkCmUAriWAXBQByIQBoYAI2FC6AelkwwIWBxIkQ5KgEIA-MxrNmoSLDgAmehiSFBMUhWp0m+Vuy69++WyPFSYIuQUyCBhVdU1qOBCOAA8ABw5gKA4AE31DY2h4AGYUDGw8G2IoCNpLRkLXTh4+Aj8AmHkYYNC1DQd4aPjE5LSaIA

**Comment by IllusionMH** [comment]
[Excess property checks](https://www.typescriptlang.org/docs/handbook/interfaces.html#excess-property-checks) are only performed on fresh object literals.

In cases `a1` and `a3` return types of functions are compared - no fresh literals therefore no excess property checks.

In case `a2` excess properties check is triggered because you return object literal.

Similar 
```ts
const obj = {a: '', b: ''};
const c1: {a: string} = obj; // no error
const c2: {a: string} = {a: '', b: ''}; // error

const a22 = ():{ a: string } => {
  const tmp = {a: '', b: ''};
  return  tmp; // no error
}
```
[Playground](https://www.typescriptlang.org/play/#code/MYewdgzgLgBAhgRgFwwBQEoYF4B8MDe8K0ATgJZgDmMAvtmprgQFAwwkCmUAriWAXBQByIQBoYAI2FC6AelkwwIWBxIkQ5KgEIA-MxrNmoSLDgAmehiSFBMUhWp0m+Vuy69++WyPFSYIuQUyCBhVdU1qOBCOAA8ABw5gKA4AE31DY2h4AGYUDGw8G2IoCNpLRkLXTh4+Aj8AmHkYYNC1DQd4aPjE5LSDDPAs8wssBhQiuxKOp0q2TNgoAFs4+gmfSWlaKvdamCW4ppaw9qpO0O6k1PSjQdgQCQArVe8xDf8ZAG4bkxhgZAFihEnDB7g8Po0FEpWuFvllgGZxrZ7FRgV5pL5pDRwU1YsAOBAQnF1AkSFAAJ7QjRAA)
